pkg/pkg.go

```diff
@@ -0,0 +1,214 @@
+	}
+
+	return &body, nil
+}
+
+func PushServicePackage(ctxt context.Context, srcTagName string, forcePush bool, adpt *adapter.Adapter, logger *log.Logger) (*api.PushResponseBody, error) {
+	srcTag, err := name.NewTag(srcTagName, name.WeakValidation)
+	if err != nil {
+		return nil, fmt.Errorf("invalid src tag format: %w", err)
+	}
+
+	path := pkgPath(nil) + "/push"
+	q := url.Values{}
+	q.Set("tag", srcTag.String())
+	q.Set("force", strconv.FormatBool(forcePush))
+	path += "?" + q.Encode()
+
+	// copy docker image
+	ref, err := name.ParseReference(srcTag.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse name reference: %s, %w", srcTag.String(), err)
+	}
+
+	img, err := daemon.Image(ref)
+	if err != nil {
+		return nil, fmt.Errorf("reading image %q: %w", ref, err)
+	}
+
+	errChan := make(chan error, 1)
+	updatesChan := make(chan v1.Update)
+	// Write
+	reader, writer := io.Pipe()
+	go func() {
+		defer func() {
+			if err := writer.Close(); err != nil {
+				fmt.Printf("failed to close pipe writer: %s\n", err)
+			}
+		}()
+		if err := tarball.Write(ref, img, writer, tarball.WithProgress(updatesChan)); err != nil {
+			errChan <- err
+		}
+	}()
+
+	go func() {
+		for update := range updatesChan {
+			if errors.Is(update.Error, io.EOF) {
+				fmt.Printf("\033[2K\rDone! %s completed out of %s\n", bytesize.New(float64(update.Complete)), bytesize.New(float64(update.Total)))
+				return
+			}
+			fmt.Printf("\033[2K\rUploading..., %s completed out of %s", bytesize.New(float64(update.Complete)), bytesize.New(float64(update.Total)))
+		}
+	}()
+
+	res, err := (*adpt).Post(ctxt, path, reader, -1, nil, logger)
+	if err != nil {
+		// error type assertion with goa ???
+		if strings.Contains(err.Error(), "already created") {
+			return nil, fmt.Errorf("tag: %s already created, use -f to force overwrite", srcTag)
+		}
+		return nil, fmt.Errorf("failed to push service package: %w", err)
+	}
+
+	if len(errChan) > 0 {
+		err := <-errChan
+		return nil, fmt.Errorf("failed to write image: %w", err)
+	}
+
+	var body api.PushResponseBody
+	if err := res.AsType(&body); err != nil {
+		return nil, fmt.Errorf("failed to decode update service response body; %w", err)
+	}
+
+	return &body, nil
+}
+
+func PullPackage(ctxt context.Context, tag string, adpt *adapter.Adapter, logger *log.Logger) error {
+	srcTag, err := name.NewTag(tag, name.WeakValidation)
+	if err != nil {
+		return fmt.Errorf("invalid src tag format: %w", err)
+	}
+	tag = srcTag.String()
+
+	path := pkgPath(nil) + "/pull"
+	q := url.Values{}
+	q.Set("tag", tag)
+	path += "?" + q.Encode()
+
+	handler := func(resp *http.Response, path string, logger *log.Logger) error {
+		if resp.StatusCode != 200 {
+			data, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return fmt.Errorf("failed to read res body: %w", err)
+			}
+			return fmt.Errorf("statusCode: %d, error: %s", resp.StatusCode, string(data))
+		}
+		data, err := io.ReadAll(resp.Body)
+		if err != nil {
```

Comment:
Is that the entire docker image with all layers? The AI service images can be VERY large, not sure what happens when the image is larger than the available memory?